reference/base/index_set_kernels.cpp

```diff
@@ -0,0 +1,191 @@
+    auto tmp_subset_begin = gko::vector<IndexType>(exec);
+    auto tmp_subset_end = gko::vector<IndexType>(exec);
+    auto tmp_subset_superset_index = gko::vector<IndexType>(exec);
+    tmp_subset_begin.push_back(tmp_indices.get_data()[0]);
+    tmp_subset_superset_index.push_back(0);
+    for (auto i = 1; i < num_indices; ++i) {
+        if ((tmp_indices.get_data()[i] ==
+             (tmp_indices.get_data()[i - 1] + 1)) ||
+            (tmp_indices.get_data()[i] == tmp_indices.get_data()[i - 1])) {
+            continue;
+        } else {
+            tmp_subset_end.push_back(tmp_indices.get_data()[i - 1] + 1);
+            tmp_subset_superset_index.push_back(
+                tmp_subset_superset_index.back() + tmp_subset_end.back() -
+                tmp_subset_begin.back());
+            if (i < num_indices) {
+                tmp_subset_begin.push_back(tmp_indices.get_data()[i]);
+            }
+        }
+    }
+    tmp_subset_end.push_back(tmp_indices.get_data()[num_indices - 1] + 1);
+    tmp_subset_superset_index.push_back(tmp_subset_superset_index.back() +
+                                        tmp_subset_end.back() -
+                                        tmp_subset_begin.back());
+
+    GKO_ASSERT(tmp_subset_begin.size() == tmp_subset_end.size());
+    GKO_ASSERT((tmp_subset_begin.size() + 1) ==
+               tmp_subset_superset_index.size());
+    *subset_begin = std::move(gko::Array<IndexType>(
+        exec, tmp_subset_begin.data(),
+        tmp_subset_begin.data() + tmp_subset_begin.size()));
+    *subset_end = std::move(
+        gko::Array<IndexType>(exec, tmp_subset_end.data(),
+                              tmp_subset_end.data() + tmp_subset_end.size()));
+    *superset_indices = std::move(gko::Array<IndexType>(
+        exec, tmp_subset_superset_index.data(),
+        tmp_subset_superset_index.data() + tmp_subset_superset_index.size()));
+}
+
+GKO_INSTANTIATE_FOR_EACH_INDEX_TYPE(GKO_DECLARE_INDEX_SET_POPULATE_KERNEL);
+
+
+template <typename IndexType>
+void global_to_local(std::shared_ptr<const DefaultExecutor> exec,
+                     const IndexType index_space_size,
+                     const Array<IndexType> *subset_begin,
+                     const Array<IndexType> *subset_end,
+                     const Array<IndexType> *superset_indices,
+                     const Array<IndexType> *global_indices,
+                     Array<IndexType> *local_indices)
+{
+    for (auto i = 0; i < global_indices->get_num_elems(); ++i) {
+        auto index = global_indices->get_const_data()[i];
+        GKO_ASSERT(index < index_space_size);
+        auto bucket =
+            std::distance(subset_begin->get_const_data(),
+                          std::upper_bound(subset_begin->get_const_data(),
+                                           subset_begin->get_const_data() +
+                                               subset_begin->get_num_elems(),
+                                           index));
```

Comment:
In practical usage (i.e. your usage in MPI), will the indices be sorted? If so, maybe an optimization/a specific function for the sorted version would make sense? You could start over from your previous location if the data was sorted or other faster algorithms.